core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,9 +1158,14 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
```

Comment:
@pengwei-li : My understanding is that there are two cases. In case 1, the controller path is created by the current ZK session. In this case, ZKCheckedEphemeral.create() should not throw ZKNodeExistException because of the following check in ZKCheckedEphemeral.GetDataCallback. Are you saying this didn't happen?
          case Code.OK =>
                if (stat.getEphemeralOwner != zkHandle.getSessionId)
                  setResult(Code.NODEEXISTS)
                else
                  setResult(Code.OK)
In case 2, the controller path is created by an older ZK session that's already expired. In this case, the ephemeral node should have been deleted by ZK server since the session is already expired. Are you saying the ephemeral node is still there after session expiration?

In any case, I want to confirm which case the patch is trying to fix.